challenger/src/serializing_challenger.rs

```diff
@@ -50,56 +50,77 @@ impl<F: PrimeField32, Inner: CanObserve<u8>> SerializingChallenger32<F, Inner> {
     }
 }
 
-impl<F, H> SerializingChallenger32<F, HashChallenger<u8, H, 32>>
+impl<F, T, H, const N: usize> SerializingChallenger32<F, T, HashChallenger<T, H, N>>
 where
     F: PrimeField32,
-    H: CryptographicHasher<u8, [u8; 32]>,
+    H: CryptographicHasher<T, [T; N]>,
+    T: Clone,
 {
-    pub fn from_hasher(initial_state: Vec<u8>, hasher: H) -> Self {
+    pub fn from_hasher(initial_state: Vec<T>, hasher: H) -> Self {
         Self::new(HashChallenger::new(initial_state, hasher))
     }
 }
 
-impl<F: PrimeField32, Inner: CanObserve<u8>> CanObserve<F> for SerializingChallenger32<F, Inner> {
+impl<F: PrimeField32, Inner: CanObserve<u8>> CanObserve<F> for SerializingChallenger32<F, u8, Inner> {
     fn observe(&mut self, value: F) {
         self.inner
             .observe_slice(&value.as_canonical_u32().to_le_bytes());
     }
 }
 
-impl<F: PrimeField32, const N: usize, Inner: CanObserve<u8>> CanObserve<Hash<F, u8, N>>
-    for SerializingChallenger32<F, Inner>
+impl<F: PrimeField32, Inner: CanObserve<u32>> CanObserve<F> for SerializingChallenger32<F, u32, Inner> {
+    fn observe(&mut self, value: F) {
+        self.inner
+            .observe(value.as_canonical_u32());
+    }
+}
+
+impl<F: PrimeField32, T, const N: usize, Inner: CanObserve<T>> CanObserve<Hash<F, T, N>>
+    for SerializingChallenger32<F, T, Inner>
 {
-    fn observe(&mut self, values: Hash<F, u8, N>) {
+    fn observe(&mut self, values: Hash<F, T, N>) {
         for value in values {
             self.inner.observe(value);
         }
     }
 }
 
-impl<F, EF, Inner> CanSample<EF> for SerializingChallenger32<F, Inner>
+impl<F, EF, Inner> CanSample<EF> for SerializingChallenger32<F, u8, Inner>
 where
     F: PrimeField32,
     EF: ExtensionField<F>,
     Inner: CanSample<u8>,
 {
     fn sample(&mut self) -> EF {
-        let modulus = F::ORDER_U64 as u32;
-        let log_size = log2_ceil_u64(F::ORDER_U64);
-        let pow_of_two_bound = (1 << log_size) - 1;
-        // Perform rejection sampling over the uniform range (0..log2_ceil(p))
-        let sample_base = |inner: &mut Inner| loop {
+        let log_size = log2_ceil_u64(F::ORDER_U64).saturating_sub(1);
+        let bound = (1 << log_size) - 1;
```

Comment:
The plonky3 sampling uses rejection sampling (probably changes since the last time you update the branch), so we should keep that in our versions as well